pkg/controllers/provisioning/scheduling/queue.go

```diff
@@ -0,0 +1,88 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduling
+
+import (
+	"sort"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/aws/karpenter/pkg/utils/resources"
+)
+
+// Queue is a queue of pods that is scheduled.  It's used to attempt to schedule pods as long as we are making progress
+// in scheduling. This is sometimes required to maintain zonal topology spreads with constrained pods, and can satisfy
+// pod affinities that occur in a batch of pods if there are enough constraints provided.
+type Queue struct {
+	pods       []*v1.Pod
+	lastPopped *v1.Pod
+	attempts   int
+}
+
+// NewQueue constructs a new queue given the input pods, sorting them to optimize for bin-packing into nodes.
+func NewQueue(pods ...*v1.Pod) *Queue {
+	sort.Slice(pods, byCPUAndMemoryDescending(pods))
```

Comment:
Yes, there are some ideas around "vector bin packing" that I want to investigate at some point that will probably improve upon this.  It will take writing some benchmark code to ensure that we are actually packing more densely, but a different sort order could improve.  

We may also want to sort pods with anti-affinity or topology spreads first.  They are more likely to create new nodes which we can then use to fulfill other scheduling constraints.  Lots of options to investigate and determine what works best.